switchtec.c

```diff
@@ -1295,6 +1431,11 @@ static int switchtec_pci_probe(struct pci_dev *pdev,
 		  &stdev->mmio_part_cfg->mrpc_comp_hdr);
 	enable_link_state_events(stdev);
 
+	if (stdev->dma_mrpc){
+		iowrite64(stdev->mrpc_output_dma, &stdev->mmio_mrpc->dma_addr);
+		iowrite32(SWITCHTEC_DMA_MRPC_EN_IRQ, &stdev->mmio_mrpc->dma_en);
+	}
```

Comment:
Also: very important -- We need to remember to unset dma_addr before we call dma_free_coherent(). We don't want the hardware writing to that memory after it's freed.

It's also not clear to me what enables using DMA transfers. You only seem to set the flag to enable interrupt requests. Does the hardware enable DMA requests once dma_addr is set or is the EN_IRQ flag mis-named?

Also, you might want to consider some of the security implications here. On Windows, any user can now write this DMA address and can use it to scribble on any kernel-space memory. That's really bad. Makes me regret letting your colleagues convince me to allow userspace access to GAS at all... At least on Linux it requires root access and the driver doesn't actually add any privilege userspace didn't already have. Someone should probably look into ensuring GAS access on Windows at least requires elevated privileges, but I'm not sure how that can be done.

Reply:
Nice capture.
For the dma mrpc new feature: four registers added: enable, address, vector and version.
where version number greater than zero, signify dma mrpc feature is supported; dma mrpc enabled by enable register instead of filling the target address to address register. So, yes, EN_IRQ flag msi-named.
More about the IRQ:
Command register format also change for this new feature.
Where BIT [15:0] keep unchanged, should feed with mrpc command index.
BIT 16, notify or not, where 0 notify, 1 not notify.
Whether the event will trig IRQ or not, is still controlled by BIT 3 of MRPC completion header reg.